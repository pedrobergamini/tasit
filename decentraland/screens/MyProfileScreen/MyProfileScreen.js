import React from "react";
import MyProfile from "@presentational/MyProfile";
import { connect } from "react-redux";
import {
  GENERATING_ACCOUNT,
  FUNDING_WITH_ETH,
  FUNDING_WITH_MANA,
  APPROVING_MARKETPLACE,
} from "@constants/AccountCreationStatus";
import { MISSING, DONE } from "@constants/ActionStatus";
import PropTypes from "prop-types";

const creationSteps = [
  {
    name: "Account created",
    creationStatus: GENERATING_ACCOUNT,
    percentage: 0.25,
  },
  {
    creationStatus: FUNDING_WITH_ETH,
    name: "Funded with ETH",
    percentage: 0.25,
  },
  {
    creationStatus: FUNDING_WITH_MANA,
    name: "Funded with MANA tokens",
    percentage: 0.25,
  },
  {
    creationStatus: APPROVING_MARKETPLACE,
    name: "Linked to marketplace",
    percentage: 0.25,
  },
];

const stepWasDone = (step, accountInfo) => {
  const { creationActions, account } = accountInfo;
  const { creationStatus } = step;

  const statusWithAction = [
    FUNDING_WITH_ETH,
    FUNDING_WITH_MANA,
    APPROVING_MARKETPLACE,
  ];

  if (creationStatus === GENERATING_ACCOUNT) {
    const isAccountCreated = account !== null;
    return isAccountCreated;
  } else if (statusWithAction.includes(creationStatus)) {
    const hasAnAction = creationActions[creationStatus] !== undefined;
    return hasAnAction;
  }
};

export class MyProfileScreen extends React.Component {
  render() {
    const { accountInfo } = this.props;

    const creationStepsWithStatus = creationSteps.map(step => {
      const wasDone = stepWasDone(step, accountInfo);
      // TODO: As soon as we store action status in redux, this logic will change
      // transaction pending, confirmed once, confirmed many times, failed, etc.
      const status = wasDone ? DONE : MISSING;
      return { ...step, status };
    });

    return (
      <MyProfile
        progress={this._getPercentage(creationStepsWithStatus)}
        creationSteps={creationStepsWithStatus}
      />
    );
  }

  _getPercentage(creationSteps) {
    const percentage = creationSteps
      .filter(step => step.status === DONE)
      .reduce((total, step) => total + step.percentage, 0);

    return percentage;
  }
}

MyProfileScreen.propTypes = {
  accountInfo: PropTypes.object,
};

const mapStateToProps = state => {
  const { accountInfo } = state;
  return { accountInfo };
};

export default connect(mapStateToProps)(MyProfileScreen);
