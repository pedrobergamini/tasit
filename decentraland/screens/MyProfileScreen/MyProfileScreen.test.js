import React from "react";
import { shallow } from "enzyme";
import { MyProfileScreen } from "./MyProfileScreen";
import {
  FUNDING_WITH_ETH,
  FUNDING_WITH_MANA,
  APPROVING_MARKETPLACE,
} from "@constants/AccountCreationStatus";

describe("MyProfileScreen", () => {
  describe("renders the component", () => {
    it("initial state (without account)", async () => {
      expect(
        shallow(
          <MyProfileScreen
            accountInfo={{
              account: null,
              creationActions: {},
            }}
          />
        )
      ).toMatchSnapshot();
    });

    it("funding with ethers", async () => {
      expect(
        shallow(
          <MyProfileScreen
            accountInfo={{
              account: {},
              creationActions: {
                [FUNDING_WITH_ETH]: {},
              },
            }}
          />
        )
      ).toMatchSnapshot();
    });

    it("funding with ethers with null action (used when action is unknown)", async () => {
      expect(
        shallow(
          <MyProfileScreen
            accountInfo={{
              account: {},
              creationActions: {
                [FUNDING_WITH_ETH]: null,
              },
            }}
          />
        )
      ).toMatchSnapshot();
    });

    it("funding with mana AND approving marketplace", async () => {
      expect(
        shallow(
          <MyProfileScreen
            accountInfo={{
              account: {},
              creationActions: {
                [FUNDING_WITH_ETH]: {},
                [FUNDING_WITH_MANA]: {},
                [APPROVING_MARKETPLACE]: {},
              },
            }}
          />
        )
      ).toMatchSnapshot();
    });
  });
});
