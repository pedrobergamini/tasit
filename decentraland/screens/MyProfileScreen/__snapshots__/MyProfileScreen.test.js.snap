// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyProfileScreen renders the component funding with ethers 1`] = `
<MyProfile
  creationSteps={
    Array [
      Object {
        "creationStatus": "GENERATING_ACCOUNT",
        "name": "Account created",
        "percentage": 0.25,
        "status": "done",
      },
      Object {
        "creationStatus": "FUNDING_WITH_ETH",
        "name": "Funded with ETH",
        "percentage": 0.25,
        "status": "done",
      },
      Object {
        "creationStatus": "FUNDING_WITH_MANA",
        "name": "Funded with MANA tokens",
        "percentage": 0.25,
        "status": "missing",
      },
      Object {
        "creationStatus": "APPROVING_MARKETPLACE",
        "name": "Linked to marketplace",
        "percentage": 0.25,
        "status": "missing",
      },
    ]
  }
  progress={0.5}
/>
`;

exports[`MyProfileScreen renders the component funding with ethers with null action (used when action is unknown) 1`] = `
<MyProfile
  creationSteps={
    Array [
      Object {
        "creationStatus": "GENERATING_ACCOUNT",
        "name": "Account created",
        "percentage": 0.25,
        "status": "done",
      },
      Object {
        "creationStatus": "FUNDING_WITH_ETH",
        "name": "Funded with ETH",
        "percentage": 0.25,
        "status": "done",
      },
      Object {
        "creationStatus": "FUNDING_WITH_MANA",
        "name": "Funded with MANA tokens",
        "percentage": 0.25,
        "status": "missing",
      },
      Object {
        "creationStatus": "APPROVING_MARKETPLACE",
        "name": "Linked to marketplace",
        "percentage": 0.25,
        "status": "missing",
      },
    ]
  }
  progress={0.5}
/>
`;

exports[`MyProfileScreen renders the component funding with mana AND approving marketplace 1`] = `
<MyProfile
  creationSteps={
    Array [
      Object {
        "creationStatus": "GENERATING_ACCOUNT",
        "name": "Account created",
        "percentage": 0.25,
        "status": "done",
      },
      Object {
        "creationStatus": "FUNDING_WITH_ETH",
        "name": "Funded with ETH",
        "percentage": 0.25,
        "status": "done",
      },
      Object {
        "creationStatus": "FUNDING_WITH_MANA",
        "name": "Funded with MANA tokens",
        "percentage": 0.25,
        "status": "done",
      },
      Object {
        "creationStatus": "APPROVING_MARKETPLACE",
        "name": "Linked to marketplace",
        "percentage": 0.25,
        "status": "done",
      },
    ]
  }
  progress={1}
/>
`;

exports[`MyProfileScreen renders the component initial state (without account) 1`] = `
<MyProfile
  creationSteps={
    Array [
      Object {
        "creationStatus": "GENERATING_ACCOUNT",
        "name": "Account created",
        "percentage": 0.25,
        "status": "missing",
      },
      Object {
        "creationStatus": "FUNDING_WITH_ETH",
        "name": "Funded with ETH",
        "percentage": 0.25,
        "status": "missing",
      },
      Object {
        "creationStatus": "FUNDING_WITH_MANA",
        "name": "Funded with MANA tokens",
        "percentage": 0.25,
        "status": "missing",
      },
      Object {
        "creationStatus": "APPROVING_MARKETPLACE",
        "name": "Linked to marketplace",
        "percentage": 0.25,
        "status": "missing",
      },
    ]
  }
  progress={0}
/>
`;
